### PR TITLE
User management

### PR DESCRIPTION
**Pull request by AnikaAgiwal2711**
# Description
A few sentences describing the overall goals of the pull request's commits.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| |

# Checklist:

- [X] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [X] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 80% code coverage
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] Backward compatibility is not broken

# How Has This Been Tested?
acceptance testing
image

image

### CODE REVIEW

**Comment by terraform-providers-svc** [comment]
Can one of the admins verify this patch?

File: docs/resources/user.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+
+
+<!-- schema generated by tfplugindocs -->
+## Schema
+
+### Required
+
+- `name` (String) The name of the user.
+- `password` (String) Password of the user.
```
Comment (rounak-adhikary) [inline docs/resources/user.md:38]: Please add in descriptions that these fields cannot be updated. All else lgtm.

Comment (AnikaAgiwal2711) [reply docs/resources/user.md:38]: Sure

File: powerflex/provider/user_resource.go [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+	// Retrieve values from plan
+	var plan models.UserModel
+
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	system, err := helper.GetFirstSystem(r.client)
```
Comment (shenda1) [inline powerflex/provider/user_resource.go:111]: Can we move this block to configure() method? https://github.com/dell/terraform-provider-powerflex/blob/main/powerflex/provider/device_resource.go#L194

File: powerflex/provider/user_resource.go [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+	diags = resp.State.Set(ctx, state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+}
+
+// UpdateUserState is the helper function that marshals API response to UserModel
+func UpdateUserState(user *scaleiotypes.User, plan models.UserModel) models.UserModel {
```
Comment (shenda1) [inline powerflex/provider/user_resource.go:151]: This can be moved to the helper package

File: powerflex/provider/user_resource.go [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+	resp.Schema = schema.Schema{
+		Description:         "This resource can be used to manage mapping of volumes to an SDC on a PowerFlex array.",
+		MarkdownDescription: "This resource can be used to manage mapping of volumes to an SDC on a PowerFlex array.",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Description:         "The ID of the user.",
+				Computed:            true,
+				MarkdownDescription: "The ID of the user.",
+			},
+			"name": schema.StringAttribute{
```
Comment (shenda1) [inline powerflex/provider/user_resource.go:61]: validation for length is needed?

File: powerflex/provider/user_resource_test.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
```
[the remark this answers is not in the document]
Comment (AnikaAgiwal2711) [reply powerflex/provider/user_resource_test.go:21]: As discussed , It panics out . So we'll be changing it later on for all the test files.

File: powerflex/provider/user_resource.go (hunk 4 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/user_resource.go:61]: added  the validation for name and password

File: powerflex/provider/user_resource.go (hunk 3 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/user_resource.go]: Moved to the helper package

File: powerflex/provider/user_resource.go (hunk 2 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/user_resource.go]: Done.

File: docs/resources/user.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+title: "powerflex_user resource"
+linkTitle: "powerflex_user"
+page_title: "powerflex_user Resource - powerflex"
+subcategory: ""
+description: |-
+  This resource can be used to manage mapping of volumes to an SDC on a PowerFlex array.
```
Comment (shenda1) [inline docs/resources/user.md:22]: Please change the description. Applicable for all instances.

File: powerflex/provider/user_resource.go [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+				Description: "The name of the user." +
```
Comment (shenda1) [inline powerflex/provider/user_resource.go:62]: I think the update part can be included in MarkDownDescription as well.

File: powerflex/provider/user_resource.go [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+					" Cannot be updated.",
+				Required: true,
+				Validators: []validator.String{
```
Comment (Krunal-Thakkar) [inline powerflex/provider/user_resource.go:65]: Max length validation?

File: powerflex/provider/user_resource.go [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+					stringvalidator.LengthAtLeast(1),
+				},
+				MarkdownDescription: "The name of the user.",
+			},
+			"role": schema.StringAttribute{
+				Description: "The role of the user." +
+					" Accepted values are 'Administrator', 'Monitor', 'Configure', 'Security', 'FrontendConfig', 'BackendConfig'.",
+				Required:            true,
+				MarkdownDescription: "The role of the user.",
+				Validators: []validator.String{stringvalidator.OneOf(
```
Comment (Krunal-Thakkar) [inline powerflex/provider/user_resource.go:75]: do we require case sensitive?

Comment (AnikaAgiwal2711) [reply powerflex/provider/user_resource.go:75]: Yeah it is case sensitive. I tried "monitor' and got this error : 
{
    "message": "userRole should get one of the following values: Monitor, Configure, Administrator, Security, FrontendConfig, BackendConfig, but its value is monitor.",
    "httpStatusCode": 400,
    "errorCode": 0
}

File: powerflex/provider/user_resource.go (hunk 7 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/user_resource.go:68]: Done

File: docs/resources/user.md (hunk 6 above)
Comment (AnikaAgiwal2711) [reply docs/resources/user.md:22]: Done

**Review by rounak-adhikary — DISMISSED** [review]
LGTM

**Review by shenda1 — DISMISSED** [review]

File: powerflex/provider/user_resource.go (hunk 8 above)
Comment (Krunal-Thakkar) [reply powerflex/provider/user_resource.go:71]: 31 character is max length for name

**Comment by shenda1** [comment]
run e2e test

**Comment by Krunal-Thakkar** [comment]
run e2e test

File: powerflex/provider/user_resource.go (hunk 8 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/user_resource.go:71]: Done. Thanks for providing the info

**Comment by AnikaAgiwal2711** [comment]
run e2e test

**Comment by Krunal-Thakkar** [comment]
run e2e test

**Review by shenda1 — APPROVED** [review]

**Review by Krunal-Thakkar — APPROVED** [review]

**Pull request merged**